tests/asgi/tests.py

```diff
@@ -194,6 +197,56 @@ async def test_post_body(self):
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Echo!")
 
+    async def test_cancel_post_request_with_sync_processing(self):
+        loop = asyncio.get_event_loop()
+        # Events used to monitor the view processing from the parent test code
+        view_started_event = asyncio.Event()
+        view_finished_event = asyncio.Event()
+        # Record exception raised in the test view
+        exceptions = []
+
+        # This view will run in a new thread because it is wrapped in sync_to_async.
+        # The view consumes the POST body data after a short delay. The test will cancel
+        # the request using http.disconnect during the delay, but because this is a sync
+        # view the code runs to completion. There should be no exceptions raised inside
+        # the view code.
+        @csrf_exempt
+        @sync_to_async
+        def echo_view(request):
+            try:
+                loop.call_soon_threadsafe(view_started_event.set)
+                time.sleep(0.1)
+                return HttpResponse(request.body)
+            except Exception as e:
+                exceptions.append(e)
+            finally:
+                loop.call_soon_threadsafe(view_finished_event.set)
+
+        # Request class to use the view.
+        class TestASGIRequest(ASGIRequest):
+            urlconf = (path("echo/", echo_view),)
+
+        # Handler to use request class.
+        class TestASGIHandler(ASGIHandler):
+            request_class = TestASGIRequest
+
+        application = TestASGIHandler()
+        scope = self.async_request_factory._base_scope(
+            method="POST",
+            path="/echo/",
+        )
+        communicator = ApplicationCommunicator(application, scope)
+
+        await communicator.send_input({"type": "http.request", "body": b"Echo!"})
+
```

Comment:
Could we assert that the response that was returned from the view actually matches `Echo!`? :thinking:

Reply:
Because we are sending the `http.disconnect` the application never actually returns the response, even though the view code runs concurrently to create it. Perhaps I could put an assertion in the view code, or capture the contents of request.body in there.

Reply:
I've updated the test so that rather than being an echo, the view just records the content of request_body which can then be asserted at the end of the test

Reply:
I like this, thanks!